FlowCrypt/src/main/java/com/flowcrypt/email/jetpack/workmanager/HandlePasswordProtectedMsgWorker.kt

```diff
@@ -0,0 +1,425 @@
+import kotlinx.coroutines.withContext
+import org.bouncycastle.openpgp.PGPPublicKeyRingCollection
+import org.bouncycastle.openpgp.PGPSecretKeyRingCollection
+import org.pgpainless.util.Passphrase
+import java.io.ByteArrayOutputStream
+import java.io.FileNotFoundException
+import java.io.InputStream
+import java.net.SocketException
+import java.util.Base64
+import java.util.Properties
+import javax.mail.Message
+import javax.mail.MessagingException
+import javax.mail.Multipart
+import javax.mail.Session
+import javax.mail.internet.InternetAddress
+import javax.mail.internet.MimeBodyPart
+import javax.mail.internet.MimeMessage
+import javax.net.ssl.SSLException
+import kotlin.random.Random
+
+/**
+ * @author Denis Bondarenko
+ *         Date: 12/29/21
+ *         Time: 9:30 AM
+ *         E-mail: DenBond7@gmail.com
+ */
+class HandlePasswordProtectedMsgWorker(context: Context, params: WorkerParameters) :
+  BaseMsgWorker(context, params) {
+  override suspend fun doWork(): Result = withContext(Dispatchers.IO) {
+    LogsUtil.d(TAG, "doWork")
+
+    if (isStopped) {
+      return@withContext Result.success()
+    }
+
+    try {
+      val account = AccountViewModel.getAccountEntityWithDecryptedInfoSuspend(
+        roomDatabase.accountDao().getActiveAccountSuspend()
+      ) ?: return@withContext Result.success()
+
+      val passwordProtectedCandidates = roomDatabase.msgDao().getOutboxMsgsByStatesSuspend(
+        account = account.email,
+        msgStates = listOf(MessageState.NEW_PASSWORD_PROTECTED.value)
+      )
+
+      if (passwordProtectedCandidates.isNotEmpty()) {
+        prepareAndUploadPasswordProtectedMsgsToFES(account)
+      }
+
+      return@withContext rescheduleIfActiveAccountWasChanged(account)
+    } catch (e: Exception) {
+      e.printStackTrace()
+      return@withContext Result.failure()
+    } finally {
+      LogsUtil.d(TAG, "work was finished")
+    }
+  }
+
+  private suspend fun prepareAndUploadPasswordProtectedMsgsToFES(account: AccountEntity) =
+    withContext(Dispatchers.IO) {
+      val apiRepository = FlowcryptApiRepository()
+      val keysStorage = KeysStorageImpl.getInstance(applicationContext)
+      val accountSecretKeys = PGPSecretKeyRingCollection(keysStorage.getPGPSecretKeyRings())
+
+      var list: List<MessageEntity>
+      var lastMsgUID = 0L
+      while (true) {
+        list = roomDatabase.msgDao().getOutboxMsgsByStatesSuspend(
+          account = account.email,
+          msgStates = listOf(MessageState.NEW_PASSWORD_PROTECTED.value)
+        )
+
+        if (list.isEmpty()) {
+          break
+        }
+
+        val iterator = list.iterator()
+        var msgEntity: MessageEntity? = null
+
+        while (iterator.hasNext()) {
+          val tempMsgDetails = iterator.next()
+          if (tempMsgDetails.uid > lastMsgUID) {
+            msgEntity = tempMsgDetails
+            break
+          }
+        }
+
+        if (msgEntity == null) {
+          msgEntity = list[0]
+        }
```

Comment:
last N lines from `if (list.isEmpty()) { ...` to here could be just 2 lines:
``` kotlin
if (list.isEmpty()) break
val msgEntity = list.firstOrNull { it.uid > lastMsgUID } ?: list[0]
```

Reply:
It's a good one. This code is old and from Java realization.